runtime/src/comm/ofi/comm-ofi.c

```diff
@@ -5446,13 +5510,87 @@ void amCheckLiveness(void) {
 // Interface: RMA
 //
 
+static inline
+void nb_handle_init(nb_handle_t h) {
+  h->id = chpl_task_getId();
+  h->reported = false;
+  atomic_init_bool(&h->complete, false);
+  h->next = NULL;
+}
+
+static inline
+void nb_handle_destroy(nb_handle_t h) {
+  atomic_destroy_bool(&h->complete);
+}
+
+/*
+ * put_prologue
+ *
+ * Common prologue operations for chpl_comm_put and chpl_comm_put_nb. Returns
+ * true if the PUT should proceed, false if it was handled in this function.
+ */
+static inline
+chpl_bool put_prologue(void* addr, c_nodeid_t node, void* raddr, size_t size,
+                       int32_t commID, int ln, int32_t fn) {
+
+  chpl_bool proceed = false;
+  retireDelayedAmDone(false /*taskIsEnding*/);
+
+  //
+  // Sanity checks, self-communication.
+  //
+  CHK_TRUE(addr != NULL);
+  CHK_TRUE(raddr != NULL);
+
+  if (size == 0) {
+    goto done;
+  }
+
+  if (node == chpl_nodeID) {
+    memmove(raddr, addr, size);
+    goto done;
+  }
```

Comment:
I know we already talked about this offline, but this kind of "goto done" pattern does not sit right with me. I understand the usefulness of having the cleanup code all done in one place that is jumped to, making it harder to forget to call the cleanup code (leading to leaks, etc). But in this case there is no cleanup, just "return proceed". 

Which to my mind suggests the following "simple fix"
1. remove the goto and inline "return proceed"
2. remove "proceed" altogether, and just make it "return false" and "return true
I think this change will improve readability.

However, this is arguably still not great and I know you expressed a strong preference for a single return point per function. So consider restructuring as
```
if size == 0
  do nothing
  proceed is false
else if node is current node
  do memove
  proceed is false
else 
  put_prologue_inner
  proceed is true
return proceed
```

Then all the logic for the prologue goes into the `_inner` method (or `_helper` or whatever descriptive name makes sense)

The same note for `check_complete`

Reply:
As a final note, if extra function overhead is a concern I would suggest just using `static inline`, or use `static ___always_inline` (defined in sys_basic.h) if you really want it inlined

Reply:
Done.